libevmasm/AssemblyItem.h

```diff
@@ -63,6 +63,9 @@ class AssemblyItem
 
 	AssemblyItem(u256 _push, langutil::SourceLocation _location = langutil::SourceLocation()):
 		AssemblyItem(Push, std::move(_push), std::move(_location)) { }
+	/// Used only for the free memory pointer as "memoryguard". Should probably be replaced by a separate AssemblyItemType.
+	AssemblyItem(std::shared_ptr<u256> _pushData, langutil::SourceLocation _location = langutil::SourceLocation()):
+		m_type(Push), m_data(std::move(_pushData)), m_location(std::move(_location)) { }
```

Comment:
Ideally I'd like to avoid introducing a dedicated ``AssemblyItem`` for this purpose... but this is quite hacky as well...
Context is: I need to be able to adjust the value of the free memory pointer later on, after it's already been pushed as assembly item - keeping a reference to the shared pointer allows that, but it's contrary to the design of these things to mainly be immutable...

Reply:
I think it's better to introduce a new assembly item type. How would you avoid the optimizer messing with it, for example?

Reply:
It will be fixed before the optimizer runs on it. I.e. it's only modified while running the yul optimizer during code generation which is before the old optimizer will ever run, respectively before the assembly item in question is ever touched again...

Reply:
A new kind of assembly item will of course work, but that's a much heavier change that arguably affects tooling and everything...

Reply:
To make it clearer what happens here:

Among the first things happening during old codegen is initializing the free-memory pointer. So old codegen emits a push assembly item with a fixed value containing the start of free memory.

Then at some point (during ``CompilerContext::appendInlineAssembly``, so still as part of code generation) it runs ``CompilerContext::optimizeYul``. Running stack-to-memory during that will want to change the free memory start. So the previously pushed value of the assembly item changes, but it changes only here, not later.

After old code generation the value will then again remain immutable, so the old optimizer is not actually affected - so in that sense we *could* live without a specialized assembly item.

But it means that we'd need to break the immutability of either ``Assembly`` (replacing the entire push assembly item) or ``AssemblyItem`` (replacing the pushed value) somewhere.

Making the push value of an ``AssemblyItem`` is trivial, since it already stores its push data as shared pointer - I'm still not sure if that's not more pragmatic than an entirely new assembly item.

@chriseth Given that, would you still say we should go for an entirely new assembly item type? [I'm asking because this stuff is a candidate for extracting from this PR next]